azkaban-common/src/main/java/azkaban/project/DirectoryYamlFlowLoader.java

```diff
@@ -222,25 +231,97 @@ private void addEdges(final AzkabanNode node, final AzkabanFlow azkabanFlow,
-      // Invalid: two_success, one_faileddd, {one_failed}
-      final String patternString =
-          "(?i)\\b(" + StringUtils.join(ConditionOnJobStatus.values(), "|") + ")\\b";
-      final Pattern pattern = Pattern.compile(patternString);
-      final Matcher matcher = pattern.matcher(condition);
-
-      // Todo jamiesjc: need to add validation for condition
-      while (matcher.find()) {
-        logger.info("Found conditionOnJobStatus: " + matcher.group(1));
+  private void validateCondition(final Node node, final AzkabanNode azkabanNode,
+      final AzkabanFlow azkabanFlow) {
+    boolean valid = true;
+    boolean foundConditionOnJobStatus = false;
+    String condition = azkabanNode.getCondition();
+    if (condition == null) {
+      return;
+    }
+    // First, remove all the whitespaces and parenthesis ().
+    final String replacedCondition = condition.replaceAll("\\s+|\\(|\\)", "");
+    // Second, split the condition by operators &&, ||, ==, !=, >, >=, <, <=
+    final String[] operands = replacedCondition.split(VALID_CONDITION_OPERATORS);
+    // Third, check whether all the operands are valid: only conditionOnJobStatus macros, numbers,
+    // strings, and variable substitution ${jobName:param} are allowed.
+    for (int i = 0; i < operands.length; i++) {
+      final Matcher matcher = CONDITION_ON_JOB_STATUS_PATTERN.matcher(operands[i]);
+      if (matcher.matches()) {
+        this.logger.info("Operand " + operands[i] + " is a condition on job status.");
+        if (foundConditionOnJobStatus) {
+          this.errors.add("Invalid condition for " + node.getId()
+              + ": cannot combine more than one conditionOnJobStatus macros.");
+          valid = false;
+        }
+        foundConditionOnJobStatus = true;
         node.setConditionOnJobStatus(ConditionOnJobStatus.fromString(matcher.group(1)));
         condition = condition.replace(matcher.group(1), "true");
-        logger.info("condition is " + condition);
-        node.setCondition(condition);
+      } else {
+        if (operands[i].startsWith("!")) {
+          // Remove the operator '!' from the operand.
+          operands[i] = operands[i].substring(1);
+        }
+        if (operands[i].equals("")) {
+          this.errors
+              .add("Invalid condition for " + node.getId() + ": operand is an empty string.");
+          valid = false;
+        } else if (!DIGIT_STRING_PATTERN.matcher(operands[i]).matches() &&
+            !validateVariableSubstitution(operands[i], azkabanNode, azkabanFlow)) {
+          valid = false;
+        }
+      }
+    }
+    if (valid) {
+      node.setCondition(condition);
+    } else {
+      this.logger.info("Condition of " + node.getId() + ": " + azkabanNode.getCondition()
+          + " is invalid, set it to false");
+      node.setCondition("false");
+    }
+  }
+
+  private boolean validateVariableSubstitution(final String operand, final AzkabanNode azkabanNode,
```

Comment:
isValidVariableSubstitution as it returns a boolean?

Reply:
Updated, thanks.